syntax = "proto3";
package protos.playbooks;

import "google/protobuf/wrappers.proto";
import "google/protobuf/struct.proto";

enum ResultType {
  UNKNOWN = 0;
  TIMESERIES = 1;
  TABLE = 2;
  API_RESPONSE = 3;
  BASH_COMMAND_OUTPUT = 4;
  TEXT = 5;
  LOGS = 6;
}

message LabelValuePair {
  google.protobuf.StringValue name = 1;
  google.protobuf.StringValue value = 2;
}

message TimeseriesResult {
  message LabeledMetricTimeseries {
    message Datapoint {
      sfixed64 timestamp = 1;
      google.protobuf.DoubleValue value = 2;
    }
    repeated LabelValuePair metric_label_values = 1;
    google.protobuf.StringValue unit = 2;
    repeated Datapoint datapoints = 3;
  }
  google.protobuf.StringValue metric_name = 1;
  google.protobuf.StringValue metric_expression = 2;
  repeated LabeledMetricTimeseries labeled_metric_timeseries = 3;
}

message TableResult {
  message TableColumn {
    google.protobuf.StringValue name = 1;
    google.protobuf.StringValue type = 2;
    google.protobuf.StringValue value = 3;
  }

  message TableRow {
    repeated TableColumn columns = 1;
  }

  google.protobuf.StringValue raw_query = 1;
  google.protobuf.UInt64Value total_count = 2;
  repeated TableRow rows = 3;
}

message ApiResponseResult {
  google.protobuf.StringValue request_method = 1;
  google.protobuf.StringValue request_url = 2;
  google.protobuf.UInt64Value response_status = 3;
  google.protobuf.Struct response_headers = 4;
  google.protobuf.Struct response_body = 5;
  google.protobuf.Struct error = 6;
  google.protobuf.Struct metadata = 7;
}

message BashCommandOutputResult {
  message CommandOutput {
    google.protobuf.StringValue command = 1;
    google.protobuf.StringValue output = 2;
  }
  repeated CommandOutput command_outputs = 1;
}

message TextResult {
  google.protobuf.StringValue output = 1;
}


message Result {
  google.protobuf.StringValue error = 1;
  ResultType type = 2;
  oneof result {
    TimeseriesResult timeseries = 101;
    TableResult table = 102;
    ApiResponseResult api_response = 103;
    BashCommandOutputResult bash_command_output = 104;
    TextResult text = 105;
    TableResult logs = 106;
  }
}